code/game/gamemodes/objective.dm

```diff
@@ -1057,16 +1058,22 @@ GLOBAL_LIST_EMPTY(possible_items_special)
 			for(var/C in 1 to ships_remaining)
 				var/datum/starship/ship_to_spawn = pickweight(spawnable_ships)
 				var/datum/starship/ship_spawned = SSship.create_ship(ship_to_spawn,faction,target_system)
-				ship_spawned.mission_ai = new /datum/ship_ai/guard
-				ship_spawned.mission_ai:assigned_system = target_system
+				ship_spawned.mission_ai = new /datum/ship_ai/guard/
+				var/datum/ship_ai/guard/AI = ship_spawned.mission_ai
+				AI.assigned_system = target_system
 				ship_spawned.boarding_chance = -1 //Stops boarding on all these ships. They don't need distracting.
+			if(current_wave == total_waves)
+				completed = TRUE //They won. They can either stay and fight or bail
 		else if(!ships_remaining && wave_active)
-			wave_active = FALSE
-			next_wave_start_time = world.time + rand(100,300)
-			current_wave++
-			update_explanation_text()
-	if(current_wave > total_waves && holding_system)
-		current_wave = total_waves
+			if(current_wave < total_waves)
+				wave_active = FALSE
+				next_wave_start_time = world.time + rand(100,300)
+				current_wave++
+				update_explanation_text()
+
+/datum/objective/ftl/hold_system/check_completion()
+	manage_waves()
+	if(completed)
```

Comment:
Don't use completed. Use if(manage_waves()) and make manage waves return TRUE or FALSE instead of changing the completed var

is this kinda dumb, yes.

but I think it'll be better for code quality to keep it in 2 procs